src/Common/src/TypeSystem/TypesDebugInfoWriter/UserDefinedTypeDescriptor.cs

```diff
@@ -2,41 +2,198 @@
     {
-        public UserDefinedTypeDescriptor(ITypesDebugInfoWriter objectWriter)
+        object _lock = new object();
+        bool _is64bit;
+        TargetAbi _abi;
+
+        public UserDefinedTypeDescriptor(ITypesDebugInfoWriter objectWriter, bool is64Bit, TargetAbi abi)
         {
             _objectWriter = objectWriter;
+            _is64bit = is64Bit;
+            _abi = abi;
         }
 
-        public uint GetVariableTypeIndex(TypeDesc type, bool needsCompleteIndex)
+        // Get type index for use as a variable/parameter
+        public uint GetVariableTypeIndex(TypeDesc type)
         {
-            uint typeIndex = 0;
+            lock (_lock)
+            {
+                return GetVariableTypeIndex(type, true);
+            }
+        }
+
+        // Get Type index for this pointer of specified type
+        public uint GetThisTypeIndex(TypeDesc type)
+        {
+            lock (_lock)
+            {
+                uint typeIndex;
+
+                if (_thisTypes.TryGetValue(type, out typeIndex))
+                    return typeIndex;
+
+                PointerTypeDescriptor descriptor = new PointerTypeDescriptor();
+                // Note the use of GetTypeIndex here instead of GetVariableTypeIndex (We need the type exactly, not a reference to the type (as would happen for arrays/classes), and not a primitive value (as would happen for primitives))
+                descriptor.ElementType = GetTypeIndex(type, true);
+                descriptor.Is64Bit = _is64bit ? 1 : 0;
+                descriptor.IsConst = 1;
+                descriptor.IsReference = 0;
+
+                typeIndex = _objectWriter.GetPointerTypeIndex(descriptor);
+                _thisTypes.Add(type, typeIndex);
+                return typeIndex;
+            }
+        }
+
+        // Get type index for method
+        public uint GetMethodTypeIndex(MethodDesc method)
+        {
+            lock (_lock)
+            {
+                uint typeIndex;
+
+                if (_methodIndices.TryGetValue(method, out typeIndex))
+                    return typeIndex;
+
+                MemberFunctionTypeDescriptor descriptor = new MemberFunctionTypeDescriptor();
+                MethodSignature signature = method.Signature;
+
+                descriptor.ReturnType = GetVariableTypeIndex(DebuggerCanonicalize(signature.ReturnType));
+                descriptor.ThisAdjust = 0;
+                descriptor.CallingConvention = 0x4; // Near fastcall
+                descriptor.TypeIndexOfThisPointer = signature.IsStatic ? (uint)PrimitiveTypeDescriptor.TYPE_ENUM.T_VOID : GetThisTypeIndex(method.OwningType);
+                descriptor.ContainingClass = GetTypeIndex(method.OwningType, true);
+
+                try
+                {
+                    descriptor.NumberOfArguments = checked((ushort)signature.Length);
+                }
+                catch (OverflowException)
+                {
+                    return 0;
```

Comment:
Is 0 a sentinel for "invalid"? The caller doesn't do anything special with this value - or does this just quietly make us show no type info in the debugger?